classes/local/object_manipulator/candidates/pusher_candidates.php

```diff
@@ -0,0 +1,91 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Class pusher_candidates
+ * @package tool_objectfs
+ * @author Gleimer Mora <gleimermora@catalyst-au.net>
+ * @copyright Catalyst IT
+ * @license http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace tool_objectfs\local\object_manipulator\candidates;
+
+use stdClass;
+use tool_objectfs\local\store\azure\client as azure_client;
+use tool_objectfs\local\store\object_client_base;
+use tool_objectfs\local\store\swift\client as swift_client;
+
+defined('MOODLE_INTERNAL') || die();
+
+require_once($CFG->dirroot . '/admin/tool/objectfs/tests/classes/test_client.php');
+
+class pusher_candidates extends manipulator_candidates_base {
+
+    /**
+     * pusher_candidates constructor.
+     * @param stdClass $config
+     */
+    public function __construct(stdClass $config) {
+        parent::__construct($config);
+        $this->config->maximumfilesize = \tool_objectfs\tests\test_client::MAX_UPLOAD;
+        if (in_array($config->filesystem, ['tool_objectfs\digitalocean_file_system', 'tool_objectfs\s3_file_system'])) {
```

Comment:
So this is definitely a code smell to have a big if else chain like this when we're using a class system, can we refactor this to just call a static method on the client class instead

Reply:
updated

Reply:
So changing this to a map doesn't really solve this in any way - the class pusher_candidates shouldn't have any knowledge at all about the different client flavours.

As we already have the filesystem in manipulator_builder->execute, i would call $filesystem->get_external_client and then pass the filesystem client into all the manipulators. That way any of the manipulators can delegate client specific details to the client itself. This should also avoid any issues you're encountering due to 

`MicrosoftAzure\Storage\Common\Internal\Resources`